// @yamdbf/core/bin/command/base/usagecontrol/Limit.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const BaseStrings_1 = require("../../../localization/BaseStrings");
const Util_1 = require("../../../util/Util");
const Command_1 = require("../../Command");
const Middleware_1 = require("../../middleware/Middleware");
const CommandDecorators = require("../../CommandDecorators");
const { using, localizable } = CommandDecorators;
class default_1 extends Command_1.Command {
    constructor() {
        super({
            name: 'limit',
            desc: 'Limit a command to the provided roles',
            usage: '<prefix>limit <command>, <role names, ...>',
            info: 'The comma after the command name -- before the role names list -- is necessary.',
            argOpts: { separator: ',' },
            callerPermissions: ['ADMINISTRATOR']
        });
    }
    async action(message, [res, commandName, ...roleNames]) {
        const command = this.client.commands.find(c => Util_1.Util.normalize(commandName) === Util_1.Util.normalize(c.name));
        if (!command)
            return this.respond(message, res(BaseStrings_1.BaseStrings.CMD_LIMIT_ERR_UNKNOWN_COMMAND, { commandName }));
        if (command.group === 'base')
            return this.respond(message, res(BaseStrings_1.BaseStrings.CMD_LIMIT_ERR_INVALID_GROUP));
        const storage = message.guild.storage;
        let limitedCommands = await storage.settings.get('limitedCommands') || {};
        let newLimit = limitedCommands[command.name] || [];
        let roles = [];
        let invalidRoles = [];
        for (const name of roleNames) {
            let foundRole = message.guild.roles.find(role => Util_1.Util.normalize(role.name) === Util_1.Util.normalize(name));
            if (!foundRole)
                invalidRoles.push(name);
            else if (foundRole && newLimit.includes(foundRole.id))
                message.channel.send(res(BaseStrings_1.BaseStrings.CMD_LIMIT_ERR_ALREADY_LIMITER, { roleName: foundRole.name, commandName: command.name }));
            else
                roles.push(foundRole);
        }
        if (invalidRoles.length > 0)
            message.channel.send(res(BaseStrings_1.BaseStrings.CMD_LIMIT_ERR_INVALID_ROLE, { invalidRoles: invalidRoles.map(r => `\`${r}\``).join(', ') }));
        if (roles.length === 0)
            return this.respond(message, res(BaseStrings_1.BaseStrings.CMD_LIMIT_ERR_NO_ROLES));
        newLimit = newLimit.concat(roles.map(role => role.id));
        limitedCommands[command.name] = newLimit;
        await storage.settings.set('limitedCommands', limitedCommands);
        return this.respond(message, res(BaseStrings_1.BaseStrings.CMD_LIMIT_SUCCESS, { roles: roles.map(r => `\`${r.name}\``).join(', '), commandName: command.name }));
    }
}
__decorate([
    using(Middleware_1.Middleware.expect({ '<command>': 'String' })),
    localizable
], default_1.prototype, "action", null);
exports.default = default_1;
