// reaction-menu/src/buttons.js
const Utils = require('./utils')
const RC = require('reaction-core')

module.exports = {
  left: {emoji: 'â¬…'},
  right: {emoji: 'âž¡'},
  up: {emoji: 'â¬†'},
  down: {emoji: 'â¬‡'},
  one: {emoji: '1âƒ£',
    callback: (user, message, data) => {
      data.menu.select(1).catch(console.error)
      message.delete().catch(console.error)
    }},
  two: {emoji: '2âƒ£',
    callback: (user, message, data) => {
      data.menu.select(2).catch(console.error)
      message.delete().catch(console.error)
    }},
  three: {emoji: '3âƒ£',
    callback: (user, message, data) => {
      data.menu.select(3).catch(console.error)
      message.delete().catch(console.error)
    }},
  four: {emoji: '4âƒ£',
    callback: (user, message, data) => {
      data.menu.select(4).catch(console.error)
      message.delete().catch(console.error)
    }},
  five: {emoji: '5âƒ£',
    callback: (user, message, data) => {
      data.menu.select(5).catch(console.error)
      message.delete().catch(console.error)
    }},
  six: {emoji: '6âƒ£',
    callback: (user, message, data) => {
      data.menu.select(6).catch(console.error)
      message.delete().catch(console.error)
    }},
  seven: {emoji: '7âƒ£',
    callback: (user, message, data) => {
      data.menu.select(7).catch(console.error)
      message.delete().catch(console.error)
    }},
  eight: {emoji: '8âƒ£',
    callback: (user, message, data) => {
      data.menu.select(8).catch(console.error)
      message.delete().catch(console.error)
    }},
  nine: {emoji: '9âƒ£',
    callback: (user, message, data) => {
      data.menu.select(9).catch(console.error)
      message.delete().catch(console.error)
    }},
  ten: {emoji: 'ðŸ”Ÿ',
    callback: (user, message, data) => {
      data.menu.select(10).catch(console.error)
      message.delete().catch(console.error)
    }},
  search: {emoji: 'ðŸ”',
    callback: (user, message, menu) => {
      if (menu.pages.length <= 10) {
        let msg = new RC.Menu('What page?')
        let b
        let btn = { }
        for (let i = 1; i <= data.menu.pages.length; i++) {
          b = module.exports[Utils.numbers[i]]
          btn.emoji = b.emoji
          btn.run = b.callback
          msg.addButtons(btn)
        }
        message.channel.sendMenu(msg).catch(console.error)
      } else {
        message.channel.send('What page?').then(async msg => {
          msg.channel.awaitMessages(m => !isNaN(m.content), { maxMatches: 1, time: 60000 })
            .then(msgs => {
              menu.select(msgs.first().content).catch(console.error)
              msg.delete().catch(console.error)
              msgs.first().delete().catch(console.error)
            })
        }).catch(console.error)
      }
    }}
}
